src/main/java/nextstep/subway/domain/Sections.java

```diff
@@ -3,23 +3,92 @@
 
 @Embeddable
 public class Sections {
-
     @OneToMany(mappedBy = "line", cascade = CascadeType.ALL)
     private List<Section> sections = new ArrayList<>();
 
     protected Sections() {
     }
 
     public void addSections(Section section) {
-        this.sections.add(section);
+        validate(section);
+        checkUpStation(section);
+        checkDownStation(section);
+        sections.add(section);
     }
 
     public List<Section> getSections() {
         return sections;
     }
+
+    public List<Station> getStations() {
+        Set<Station> stations = new HashSet<>();
+
+        sections.forEach(section -> {
+            stations.add(section.getUpStation());
+            stations.add(section.getDownStation());
+        });
+
+        return new ArrayList<>(stations);
+    }
+
+    private void validate(Section section) {
+        if (sections.contains(section)) {
+            throw new IllegalArgumentException(DUPLICATE_SECTION_ERROR);
+        }
+
+        if (getStations().size() > 0 && !isContainsStation(section.getUpStation(), section.getDownStation())) {
+            throw new IllegalArgumentException(NOT_MATCH_STATION_ERROR);
+        }
+    }
+
+    private boolean isContainsStation(Station upStation, Station downStation) {
+        return getStations().contains(upStation) || getStations().contains(downStation);
+    }
+
+    private void checkUpStation(Section section) {
+        Station upStation = section.getUpStation();
+
+        if (matchUpStation(upStation)) {
+            Section existingUpStation = existingUpStation(upStation);
+            existingUpStation.updateUpStation(section.getDownStation(), section.getDistance());
+        }
+    }
+
+    private void checkDownStation(Section section) {
+        Station downStation = section.getDownStation();
+
+        if (matchDownStation(downStation)) {
+            Section existingUpStation = existingDownStation(downStation);
+            existingUpStation.updateDownStation(section.getUpStation(), section.getDistance());
+        }
+    }
+
+    private boolean matchUpStation(Station station) {
+        return sections.stream().anyMatch(section -> section.isEqualsUpStation(station));
+    }
+
+    private Section existingUpStation(Station station) {
+        return sections.stream()
+                .filter(section -> section.isEqualsUpStation(station))
+                .findFirst()
+                .orElseThrow(() -> new NoSuchElementException("상행역을 찾을 수 없습니다."));
+    }
+
+    private boolean matchDownStation(Station station) {
+        return sections.stream().anyMatch(section -> section.isEqualsDownStation(station));
+    }
+
+    private Section existingDownStation(Station station) {
+        return sections.stream()
+                .filter(section -> section.isEqualsDownStation(station))
+                .findFirst()
+                .orElseThrow(() -> new NoSuchElementException("하행역을 찾을 수 없습니다."));
+    }
```

Comment:
existingDownStation() 메서드는 existingUpStation()와 똑같은 행위를 수행하고 있는 메서드 같아요!
중복코드를 하나로 합치는 것도 좋을 것 같아요 :) 
여러 방법이 있겠지만, 예외 메시지만 다르므로 빈 객체를 반환하고 호출부에서 예외 메시지를 컨트롤 하는 방법도 좋아보이는데
어떻게 생각하시나요~?
```suggestion
    private Section existingStation(Station station) {
        return sections.stream()
                .filter(section -> section.isEqualsDownStation(station))
                .findFirst()
                .orElseGet(Section::new);
    }
```

Reply:
exist- 로 시작하는 메서드 명명은 찾으려는 값이 존재하는지에 대해 체크하는 boolean 형 메서드로 유추가 되는 것 같아요.
찾은 Section을 반환해준다 라는 기능을 보았을 때, findStation 과 같은 메서드 명명도 어떨까요 :) ?

Reply:
넵! 좋은 관점인것 같습니다 감사합니다 :)

Section 형태를 반환해주고 있기때문에 find가 적합해보이네요! 수정해보겠습니다 :)

Reply:
내부에서는 isEqualsDownStation\< 과 같은 하나의 다른 행위를 하고 있어서 분리한것이라서.. 합치기가 가능할지가 .. 고민되네요!  
  
혹시 어떻게 생각하시나요? @ohtaeg

Reply:
앗, 이부분은 step4 PR때 같이 이야기해보는게 어떨까 싶습니다 :)

Reply:
다음 PR에서 같의 의견을 나눠보아요 :D